### PR TITLE
html5.css & epub.css: better conformance to the HTML Standard suggested rendering

### LINKED ISSUE #521: Unusual placement of list elements
KOReader seems to place the list elements in a strange way.  Most browsers and book readers I'm aware of make sure that the outside markers are visible by either setting margin or padding of either the ul/ol or li elements.  If all these are set to zero, the markers are off the screen.  It is not, however, so in KOReader.  Could this be made consistent with other players?

Attached are a simple epub file and screenshots made in KOReader and Calibre Book Viewer.

Any support would be appreciated.

  --Gabor

NB: the zip file is the epub file, just renamed, to allow uploading.

[test.epub](https://github.com/koreader/crengine/files/11725687/test.zip)
Screenshot of Calibre Book Viewer
Screenshot of KORader

### LINKED ISSUE #551: HTML5.css, the CSS cascade, and other stylesheet questions
I'd like to update our html5.css to conform to the current HTML specs:
https://html.spec.whatwg.org/multipage/rendering.html
(I initially based it on this very old https://github.com/FriendsOfEpub/WillThatBeOverriden/blob/master/ReadingSystems/html5/html5.css , which has a lots of old rare stuff now removed)
I also would like to find a solution to properly and consistently handle HTML attributes - cf https://github.com/buggins/coolreader/pull/214#issuecomment-755952995.

So, writing all the following to get confirmation that:
1) I understand the specs and how things should work right.
2) the problems that I see we have in crengine are real problems
3) the solutions I have in mind could be good enough

So, the HTML specs do not give a full user agent stylesheet, but there are snippets in the "rendering.html" page that I could just copy and concatenate. (There are also writings, for things that can't translate easily to CSS, about how stuff should work, more on that later...)

In their writing, they speak about [presentational hints](https://html.spec.whatwg.org/multipage/rendering.html#the-css-user-agent-style-sheet-and-presentational-hints).
I've always read "presentational" somehow as "optional" :) and was fine not having them or having only parts of them... But no, it means:

> Some rules are intended for the author-level zero-specificity presentational hints part of the CSS cascade; these are explicitly called out as presentational hints.

And https://www.w3.org/TR/css-cascade-3/#preshint (striking out the alternative that is not the one decided by HTML):

> The UA may choose to honor presentational hints in a source documents markup, for example the bgcolor attribute or s element in [HTML]. All document language-based styling must be translated to corresponding CSS rules and ~~either enter the cascade as UA-origin rules or~~ be treated as author-origin rules with a specificity of zero placed at the start of the author style sheet.
Note: ~~Presentational hints entering the cascade as UA-origin rules can be overridden by author-origin or user-origin styles.~~ Presentational hints entering the cascade as author-origin rules can be overridden by author-origin styles, but not by non-important user-origin styles. Host languages should choose the appropriate origin for presentational hints with these considerations in mind

I initially thought I could just get away with them by using `table[align=left i] { --cr-hint: zero-specificity; float: left; }` and have crengine put them at the start of the selector chain, but it looks like it's not the way this should work :)
I guess they should just be after all the user-agent & user selectors - so before the authors/publisher stylesheet ones.
And they may usually be just there, as they often/always use attribute selectors [ie. `table[align]`), so get a higher specificity than the normal user agent selectors (usually just `table {}`), so they end up after them.

But re-reading about the CSS cascade: https://www.w3.org/TR/css-cascade-3/#cascading, I see we have other problems :/
Can you confirm that I read that right: each of these 8 "stylesheets" should all be applied independently, and that a lower-specificity in the author stylesheet should be applied AFTER a higher specificity in the user-agent stylesheet - that is, the ordering of specificity is done independantly in each of these 8 stylesheets/origins. (Also see below, my point about selector and declarations and `!important`...)

I can confirm that with Firefox 
It has in its [UA stylesheet](https://hg.mozilla.org/mozilla-central/raw-file/tip/layout/style/res/html.css):
`table[align="right"] {   float: right; }`
and if I put a lower specificity in my HTML `<style>`:
`table { float: none; }` or even lower: `* { float: none; }`
my `<table align="right" ...>` does not float any longer.

This does not happen with KOReader/crengine.
So, **the problem** we have is that we have only one selectors chain, and when parsing authors (publishers) stylesheets, we insert these new selectors into the main (user-agent+user tweaks) stylesheet, and we insert them by ensuring specificity ordering as if it was all a single origin.
So, we end up placing `* { float: none; }` (so it is applied first) before `table[align="right"] {   float: right; }` (so it ends up being applied, while it shouldn't).

In crengine, there is some gymnastic when we process authors stylesheets:
- we have the main stylesheet with the useragent+usertweaks rules
- when meeting a new DocFragment with document stylesheet, we stylesheet.push() to keep a clean copy of the useragent+usertweaks one
- we then merge the authors stylesheets into that chain
- and use that single chain to apply styles to nodes
- when done with a DocFragment, we stylesheet.pop() to restore the genuine useragent+usertweaks stylesheet
- and re-do all that with next DocFragments (as they mave have a different set of authors stylesheets).

**Possible solutions**:

- keep the above logic (even if uneededly expensive), but give authors stylesheets selectors a big bias to their specificity (ie. setting a high bit of the LUint32 specificity), so they end up placed after all the useragent selectors in the single selector chain.
- have 2 selector chains in our stylesheet object: one for useragent selectors, and one that we reset and clear at each DocFragment (so less work in the push()/merge()/pop() sequence), and when meeting a node, apply the first chain, and then the second chain.

Do each sound like it would work and be enough?

About the whole cascade origin stuff, for us it could be simplified:
~~1 Transition declarations [css-transitions-1]~~ Not applicable for us
~~2 Important user agent declarations~~ There isn't any in the specs, and we shouldn't need any (or if any need of that kind, it's usually hardcoded and enforced in the code).
3 Important user declarations
4 Important author declarations
~~5 Animation declarations [css-animations-1]~~ Not applicable for us
6 Normal author declarations
7 Normal user declarations
8 Normal user agent declarations

It's a bit confusing, as this talks about declarations, and not selectors. And selectors may have declarations with both normal properties and `!important` properties....
So, I'm not really sure it says what I thought about selectors chains above :/

Anyway, assuming I'm somehow not too in space, and selectors can still be ordered as I wrote above:
We actually do ensure (3) not overridden by (4) with our !important and higher_importance bits that we associate to each property on each style.
The (6) not overridden by (7+8) would be ensure by the solutions outlined above.

There is still (7) vs (8) that we may not ensure, because they both end up in the useragent stylesheet, but I guess that can be ok. The epub.css and html5.css are quite generic/low-specificity, and a user tweaking style tweaks can always help himself solving any issue.

### PR DESCRIPTION
**Pull request by poire-z**
Except for the few first commits, the following commits aimed at getting nearer to what https://html.spec.whatwg.org/multipage/rendering.html describes (except for the fancy HTML widgets we don't support like `<dialog>`, `<meter>`, `<progress>`...)
This brings a lot of little changes, so possibly new bugs - to be bumped after next KOReader stable release, and tested for many weeks before the next one.

#### In-page footnotes: better handle duplicated ids

Follow-up to last commit of #554 dd5cf73 to better support books with id= values duplicated, which can happen in Wikipedia EPUBs.
Ie.
```html
<li id="link1">txt1, see <span id="foo">foo</span></li>
<li id="link2">txt2, see <span id="foo">foo</span></li>
```
would cause `txt2` to be shown as in-page footnote for links targetting `#link1`.
This is still a bit clumsy, our way to handle multiple ids is probably not the best, but it does not require changes elsewhere. This may be revisited in the future.

#### lvrend: handle in-page footnotes in table `<caption>`

In Wikipedia EPUBs, `<caption>` are used as container of the legend below an image, and may be quite verbose and contain link to footnotes.

#### LVString: ignore CJK chars in `lStr_findWordBounds()`

Otherwise, some words composed of both CJK chars and latin chars could be returned and fed to `hyphenate()` (this `lStr_findWordBounds()` is only used for this), which would properly states some hyphenation after a CJK char is allowed, resulting in spurious hyphens drawn after a CJK char.
See https://github.com/koreader/koreader/issues/11478#issuecomment-1951424809.
Should allow closing https://github.com/koreader/koreader/issues/11478.

#### lvtext: AddLine(): handle some CJK + space edge case

Avoid flexible CJK chars not getting flexed at end of line if followed by some ASCII space(s).
See https://github.com/koreader/koreader/issues/11478#issuecomment-1951424809.

#### EPUB: look for EPUB3 cover even when EPUB2 cover advertized

With EPUB3 documents, prefer finding an EPUB3 cover, and only when none found fallback to using any specified EPUB2 cover.
See around https://github.com/koreader/koreader/pull/11491#issuecomment-1963021880, after a long investigation and time wasted.
(Another patch to exctract an image from the first DocFragment, that feels not urgently needed, is available at https://github.com/koreader/koreader/pull/11491#issuecomment-1962947041 if we ever need it.)

#### List items: proper per-specs positionning and sizing

The sweet historical crengine way of rendering list items was to have each list-item carry its marker in its own border box (so that they do not overlap with anything on the left), but it is not per-CSS-specs: markers should be drawn outside the list item border box (in its margin, or in its parent padding, and over any element on the left or in the page margin, possibly truncated, if these margin and padding do not provide sufficient room for the marker).

We can still get something similar to that historical sweet behaviour by using, on the list-item parent container (normally an OL or UL): `padding-left: -cr-special`, which will compute to the widest list items marker width.
We can then use in our user-agent stylsheet:
```css
  ol, ul { padding-left: -cr-special; }
```
instead of the absurd per-specs:
```css
  ol, ul { padding-left: 40px; }
```
(an absolute length, which will look different depending on the font size or the DPI), that any publisher caring about list item rendering would be obliged to override; if they haven't, this `-cr-special` will do better than `40px`.
See https://github.com/koreader/crengine/issues/521#issuecomment-1952341201.
Closes #521.

#### epub.css: add/use `"@media (-cr-max-cre-dom-version: 20180527)"`

Add this private `@media` keyword to replace the more verbose `-cr-ignore-if-dom-version-greater-or-equal: 20180528`.

#### fb2def.h: add more HTML element and attributes names

As they could be mentionned in stylesheet, having them known with a fixed id could avoid some stylesheet hash mismatches and spurious renderings.

#### CSS: generic support for handling presentational hints

via `-cr-hint: presentational-hint`
Initially discussed in https://github.com/koreader/crengine/issues/551#issuecomment-1890975524.
Closes #551.

#### CSS: add support for private `-cr-apply-func:`

Will help ensuring bits from the HTML specs that can't fully be expressed with CSS.
To be used in html5.css, updated in a later commit.
Many rendering issues with our lack of support for these noticed in https://github.com/koreader/koreader/issues/10770.

#### `ldomDocumentWriterFilter`: remove attribute to CSS conversion

If was only done with HTML documents (not with EPUBs) and, by translating them to a style= attribute, was giving them a too high specificity, while they should have a very low one so they are overridden by any author CSS.
Referencing old issues where I thought about doing that:
https://github.com/koreader/crengine/issues/480#issuecomment-1134357893
https://github.com/buggins/coolreader/pull/214#issuecomment-755952995

#### lvrend: more proper rendering of block images

When such images have borders, the border was taking the full page width... This commit makes the border sticks to the image (+ its padding if any).
Also make sure there is no strut, causing some blank below the image running until the border.
Make the existing addition of an image title/suptitle/subtitle attributes values as text only with FB2/FB3 documents, as nowhere in the HTML/EPUB specs it is said this should be done. See https://github.com/koreader/koreader/issues/11173#issuecomment-1950277100.

#### lvrend: keep `margin_left/right` updated when `auto`

When handling `margin: auto`, or avoiding any width overflow, we need to keep `margin_left` and `_right` updated with their used values, as they are still used later, in particular when rendering erm_final to compute the floats footprint.
(Not doing that and keeping them at 0 could have the floats footprint "eat" more into the width available for rendering text, making it taller than needed.)

#### CSS: add support for handling HTML's `align=` attribute

Will allow us to use:
```css
  center { text-align: -cr-html-align-center }
  div[align=right] { text-align: -cr-html-align-right }
```
to properly handle this (deprecated) attribute that should still be supported by browsers according to the current HTML specs: https://html.spec.whatwg.org/multipage/rendering.html#align-descendants
We implement this the exact same way as Firefox and Webkit (with `text-align:` private keywords), even if it feels like not the way to implement how it is described in the HTML specs...

#### lvrend: fix HR and images positionning when floats involved

Rework 524e502d (where we did it for `<HR>`, possibly not the right way).
Very tedious to understand how this works in browsers, tried different things until I luckily got the same rendering with various combinations of floats, margin fixed or auto, and HTML `align=`... but it's quite possible this will fail in other contexts or combinations...
Note that Firefox and Edge do handle some cases differently, so this is quite gray territory...

#### epub.css, html5.css: minor updates for easier stylesheet switch

Make epub.css and html5.css be similar with all `display:` and `white-space:`, to avoid the "_Styles have changed in such a way that fully reloading the document may be needed for a correct rendering_" popup when switching between them.
Some values are updated to be what they will be in the major update to html5.css coming in next commit.

#### epub.css, html5.css: major updates for better conformance

Thanks to all the previous commits, we can now get html5.css to ensure most of the HTML Standard suggested rendering.
Have epub.css extend html5.css (so, it benefits from all the fixes), and update only what needs to be different for a more book like look.
See https://github.com/koreader/koreader/issues/11494#issuecomment-1963045594
Should allow closing https://github.com/koreader/koreader/issues/11494.
Should allow closing https://github.com/koreader/koreader/issues/10770.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/555)

### CODE REVIEW

File: crengine/include/fb2def.h [hunk 1]
```diff
@@ -225,23 +229,29 @@ XS_TAG1I( u )
                 // (in HTML5, when mis-nested tags would close one of these,
                 // they are re-opened when leaving the mis-nested tag container)
 
+XS_TAG1I( abbr )
```
Comment (Frenzie) [inline crengine/include/fb2def.h:232]: Unlike, progress I find it surprising we didn't already have that one.

**Review by Frenzie — APPROVED** [review]
This is intended to go into the release or after?

**Comment by poire-z** [comment]
#### AFTER :)

> This brings a lot of little changes, so possibly new bugs - to be bumped after next KOReader stable release, and tested for many weeks before the next one.

I just found some typo in my new html5.css :/
If you have some tools to check a stylesheet for errors (that would not fail to bad on my `-cr-xyz` added stuff, it would be nice to feed them to it to see if there are others.

**Comment by poire-z** [comment]
I realized I have `<a>` not styled at all html5.css (old and new).
I had and have:
```css
/* crengine-: not supported
:link { color: #0000EE; }
:visited { color: #551A8B; }
:link:active, :visited:active { color: #FF0000; }
:link, :visited { text-decoration: underline; cursor: pointer; }
:focus-visible { outline: auto; }
*/
```
But I guess I should just replace this with:
`a { color: #0000EE; text-decoration: underline; cursor: pointer; }`
Shouldn't I ?

In epub.css, we had and have:
```css
/* Make links noticable */
a {
  text-decoration: underline;
  color: gray;
}
```
I don't really like having this gray color for links (see screenshots in https://github.com/koreader/koreader/issues/11503, where links themselves and the footnote number (a back link) are in light gray, hardly graspable to me.
Although it's probably more distinguishable than blue on our eInk. But do we need this distinction? Should we use a darker gray ?

**Comment by Frenzie** [comment]
> If you have some tools to check a stylesheet for errors (that would not fail to bad on my `-cr-xyz` added stuff, it would be nice to feed them to it to see if there are others.

I'll check if stylelint can be tamed sufficiently to not be overly in the way.

**Comment by poire-z** [comment]
Of course, it wouldn't detect the typo I had (missing comma after h1) as it is still some valid selector:
```css
:is(article, aside, nav, section) :is(article, aside, nav, section) :is(article, aside, nav, section) :is(article, aside, nav, section) h1
h5 {
```

**Comment by Frenzie** [comment]
> Of course, it wouldn't detect the typo I had (missing comma after h1) as it is still some valid selector:

Heh, I forgot to ask which typo you were talking about.

On the plus side :laughing:, thanks to all that effort I did fix the redundant double margin-top and margin-bottom in fb2.css.

I'm not sure if `no-descending-specificity` should remain enabled as a warning; most of it comes across as noise. I would however prefer to fix the `length-zero-no-unit` warnings and turn them into errors.

**Comment by poire-z** [comment]
Feel (you or @hius07) welcome to cleanup that ugly looking fb2.css - no style change needed, but it could be made more readable, with proper formatting, some reordering maybe, and additional comments if needed.

**Comment by Frenzie** [comment]
If you run stylelint with ` --fix` it'll automatically apply fixes btw. At the very least for the `length-zero-no-unit` rule that'd work very well.

**Comment by poire-z** [comment]
Self-ish not: squash last commit into previous before merging.

**Comment by Frenzie** [comment]
The reorder is a performance optimization?

**Comment by poire-z** [comment]
No, no performance change, in any way.
It's just for when we use "Show matched rules":
image

Showing the CSS snippets matching the rare `listing` will show (we just parse and cut into the CSS stylesheet strings the bits that do match) the listing, and the selectors that follow, which are a bit of noise, in yellow - but it is ok to get them on rare elements:
image

While for a `p`, it will show the same snippet, but will cut it just before the `p {`:
image

Without the reorder, with the previous:
`blockquote, figure, listing, p, plaintext, pre, xmp {`
on any P we look at, we would see the noisier:
image

**Comment by Frenzie** [comment]
Not showing the full selector seems potentially a bit confusing to me; I wasn't aware of that behavior. I mean in case I wanted to edit it or something.

**Comment by poire-z** [comment]
We show the full selector (singular) that matches, plus any others (that don't match) on its right (if I could easily hide them, I would).
We don't show the other selectors on the left (that did not match).
What you'll want to edit/tweak/override after using this feature is this single full selector.
Combining these selectors with commas is just to avoid duplicating the declaration, but you're usually interested only in one. (Unless you really want to see all the selectors that got `color:yellow` to fix them before you actually see their effect on the book you're reading, but this must be super rare :))
One can see the full stylesheet with all selectors with the `View xyz.css` at the bottom of the View HTML widget.

**Comment by Frenzie** [comment]
> Combining these selectors with commas is just to avoid duplicating the declaration, but you're usually interested only in one. (Unless you really want to see all the selectors that got `color:yellow` to fix them before you actually see their effect on the book you're reading, but this must be super rare :))

In the context of EPUBs specifically (as opposed to HTML+CSS in general) it is super rare for me in the sense that I seldom override anything (give or take style tweaks like classic classname footnotes), but when I do I normally do indeed want to override all of them. In your example too, if I don't like yellow here then why would I like it there. 

But there are small screen restrictions at play too… :-)

**Comment by poire-z** [comment]
I think I'll do some reorg in html5.css.
I wanted to stay as faithful and litteral with https://html.spec.whatwg.org/multipage/rendering.html and keep the snippets as they are written, but it itches me to see the following. It shouldnt cost that much to check 3 or 5 selectors, but still... and it takes room in "View matched rules".
I wanted to stay litteral for easier pickups of future changes, but that page is a mess and it will be hard to follow changes - and I guess it's not something that should ever change (except for added new fancy html elements).

3 plain selector+declarations for ol and ul:
https://github.com/koreader/crengine/blob/54ec2811366ed8d5f79a050db281f7a5d72d2d5f/cr3gui/data/html5.css#L430-L482
5 plain selector+declarations for th !
https://github.com/koreader/crengine/blob/54ec2811366ed8d5f79a050db281f7a5d72d2d5f/cr3gui/data/html5.css#L586-L636

**Comment by Frenzie** [comment]
> 5 plain selector+declarations for th !

Right, that's why I said the Stylelint output *mostly* seemed like noise, and that's the kind of stuff it's actually trying to complain about. I guess in large teams you might need such overly strict slightly misguided rules.

**Comment by poire-z** [comment]
No more `no-descending-specificity` with using `:where` (0 added specificity) instead of `:is` (specificity of the most specific inner item).
It looks to me this is alright, and that it's now just the ordering that will still make them match before others.
What do you think?
(we don't really have to fight these warnings, but if it's painless without side effects, why not.)

I realize I have to add these ugly 
```css
:where(article, aside, nav, section) :where(article, aside, nav, section) :where(article, aside, nav, section) h1,
h4 {
```
also in epub.css, whether I use :is() or :where(), otherwise:
- if using :is(), a section/section/section/h1 would use the style of h4 from html5.css (while we want the style of h4 from epub.css)
- is using :where(),  a section/section/section/h1 would use the style of h1 from epub.css (otherwise, all h1, no matter their nesting in sections, would look like a normal h1.

**Comment by Frenzie** [comment]
If you prefer it that way, why not. :-)

**Comment by poire-z** [comment]
Regarding, from your suggestion at https://github.com/koreader/crengine/pull/300#issuecomment-511154956:
```css
/* crengine+: make a new page on H1 and section, but not in EPUBs (publishers may not
 * expect it; most EPUB renderers only break page on a new DocFragment). */
section, h1 {
  -cr-only-if: -epub-document; /* only if NOT EPUB document */
    page-break-before: always;
}
section > h1:first-of-type, section section {
  -cr-only-if: -epub-document; /* only if NOT EPUB document */
    page-break-before: auto;
}
```
Should we really keep it? In html5.css (the standard rendering, which doesn't specify anything about page breaks), or only in epub.css (our book-alike stylesheet). If wouldn't have any effect on EPUBs as stated, but may on HTML or other documents translated to HTML like DocX or OTF
Also, it feels bogus, it would break on the 2nd++ H1 inside a `<section><section><section>` where they would act like a H4.
I'd rather remove them than scratched my head to find out something better, than may still do unwelcome things. (We have styletweaks to break on headings, but those to prevent them are clumsy, and actually restores the ones from epub.css...)

**Comment by Frenzie** [comment]
I'm not opposed to removing it altogether, but I think your comment really just applies to the :first-of-type — as you edited in while I was away for a few minutes.

**Comment by poire-z** [comment]
Not edited, it's the first snippet above - I just added another snippet at the bottom.
(And what "your comment really just applies" :/ I put more than one comment there :))

**Comment by Frenzie** [comment]
You already linked to my opinion — I think it would generally improve display to treat top-level section as a page-break and not for anything lower. That's what's shown in your second edited snippet, which I believe corresponds to what I originally suggested. The first strikes me as needlessly complex with the first-of-type.

But as stated I'm not opposed to removing them if you'd rather have them gone.

**Comment by Frenzie** [comment]
If you've fixed the no unit after 0 thing, you can remove this line:

https://github.com/koreader/crengine/blob/f0ba598e1773bd41bdc7450fa7f0357b212cdb23/.stylelintrc.json#L7

**Comment by poire-z** [comment]
(Added 2 commits, at the top:
`In-page footnotes: better handle duplicated ids`
`lvrend: handle in-page footnotes in table <caption>`
)

**Review by Frenzie — APPROVED** [review]
The footnotes stuff sounds good :+1:
